sdk/core/core-lro/docs/LROEngine.md

```diff
@@ -0,0 +1,213 @@
+```
+
+Note that `path` is optional here because `resourceLocation` is also optional. Here is the implementation for `@azure/core-http`:
+
+```ts
+  public async retrieveAzureAsyncResource(
+    path?: string
+  ): Promise<LROStatus<T>> {
+  const updatedArgs = { ...this.args };
+  if (updatedArgs.options) {
+    (updatedArgs.options as any).shouldDeserialize = true;
+  }
+  const { flatResponse, rawResponse } = await sendOperation(updatedArgs, {
+    ...this.spec,
+	httpMethod: "GET",
+ 	path: path)
+  });
+  return { flatResponse, rawResponse, done: true };
+```
+
+### `LROPoller`
+
+I propose to make this class also exported by `@azure/core-lro`. This class has the following type signature:
+
+```ts
+class LROPoller<TResult, TState extends PollOperationState<TResult>> extends Poller<TState, TResult>
+```
+
+so a client can instantiate the state of the polling operation with a custom type that extends the standard `PollOperationState`. This flexibility can open the door for this class to be used in Track 2 packages where it is common for LROs to have other interesting information as part of their state and customers might want to have access to it, e.g. the last time the operation was updated.
+
+The class also has the following constructor:
+
+```ts
+constructor(options: LROPollerOptions, lro: LRO<TResult>);
+```
+
+Currently `options` has `intervalInMs` to control the polling interval and `resumeFrom` to enable resuming from a serialized state. If there are new arguments to be added to the class, they could be added to the options type. For instance, we might decide to enable customers to provide a custom LRO implementation in the form of a custom [`update`](https://github.com/Azure/azure-sdk-for-js/blob/2d2c6561cac330b8720763db88705fad4e867bda/sdk/core/core-lro/src/pollOperation.ts#L63) method, such a method could be part of the options.
+
+### `coreClientLRO` and `coreHttpLRO`
+
+I propose to make these classes auto-generated by Autorest. These classes implement `LRO<T>` and various bits of `coreHttpLRO` have been showed earlier. These classes will need access to a few pieces: operation specification and operation arguments and a primitive function that can take them as input to send a request and also converts the response into one of type `LROResponse<T>` which has both the flattened and the raw responses. Furthermore, those classes can also, optionally, take as input an object of type `FinalStateVia`, which could determine where to find the results of the LRO after the operation finished. Typically, Autorest figures out the value for `FinalStateVia` from the `x-ms-long-running-operation-options` swagger extension.
+
+## Usage examples
+
+### Create an object of `coreClientLro`
+
+```ts
+const directSendOperation = async (
+  args: coreClient.OperationArguments,
+  spec: coreClient.OperationSpec
+): Promise<unknown> => {
+  return this.client.sendOperationRequest(args, spec);
```

Comment:
where does `this` come from?

Reply:
it refers to the object that has this operation, this is a rough code snippet copied from [here](https://github.com/Azure/autorest.typescript/blob/5f7acf24417c9c4b8bdac7f59726792d90c537f6/test/integration/generated/lro/src/operations/lROs.ts#L194).

Reply:
Now that I think about it, we should probably pull that into a function or something instead of inlining it in every operation like this.

Reply:
I opened https://github.com/Azure/autorest.typescript/issues/1079.